### DISCUSSION TITLE
Import Status Spec

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#345] [Maintainer discussions]
The specs related to issue #296  are ready for review. Please take a look at the attached document and follow the [review guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines).

## Document Link
https://wiki.mathesar.org/en/design/specs/import-status

@kgodey @pavish, your feedback is required.
The deadline for feedback is set for Tuesday next week, first review round will be completed by Friday 9th. Please plan your review so that at least one round of changes can be made and reviewed if needed.

**Comment by kgodey** [comment]
@ghislaineguerin The spec looks good to me. The only feedback I have is that I think it would be good to have the loading spinner next to Unnamed Table in the sidebar as well. I'm not sure we need the asterisk if we have the loading spinner.

Also general feedback on the review process: could you provide a deadline for the first review in addition to the deadline for the final review to be completed? That way we don't have to guess how long one round of changes will need.

**Reply by ghislaineguerin** [reply]
@kgodey good point re: review rounds. I will update both.

**Reply by ghislaineguerin** [reply]
@kgodey the asterisk is for the unsaved state so I'm not sure we want to remove it until the table is saved.

**Comment by pavish** [comment]
@ghislaineguerin I have few concerns and I am unclear of the behaviour with regards to the following:
* Consider a user uploads file for Table 'A' and then moved to Table 'B' and starts editing 'B'. We show the status pane. Now the upload for 'A' completes. Do we close the pane on our own, or do we show a tick icon mentioning that the upload is complete?
  * If we close the pane on our own, it will cause the table to increase/decrease in size, which might affect user actions on the table i.e if the user is editing the table.
  * If we do not close the pane on our own, the user will have to visit the tab where upload is happening. Only then the indicator clears. We need this behaviour to be more detailed on the spec.

* The status pane covers an entire row. It takes additional space, which the table could possibly consume. The user cannot hide it entirely. I think an option for the user to show/hide it entirely is required.
* There might be additional async operations in the future. If we have a generic pane, we can show it all on one place, instead of the pane being specific to import.
* If we can turn this into a notification style option on the top bar, it might be more convenient to the user. **The user may upload on one DB, and then move to a different DB. The upload won't be cancelled, and we still have to show the indication somewhere.**

**Reply by kgodey** [reply]
I like the idea of having a general area to notify the user about things that are happening asynchronously. The notification style option in the top bar seems like it might be more familiar to users from other applications as well.

**Comment by ghislaineguerin** [comment]
@pavish @kgodey I've updated the prototype to show the status indicator on the top bar. However, I do have some questions about how we want to handle multiple vs. single processes:

![image](https://user-images.githubusercontent.com/845767/125256196-840ccf00-e2fc-11eb-8731-3853ea3a68d9.png)

For example, if we have a single file, could we show that file name and only use a count and dropdown when we have more than 1? This could save additional clicks when navigating back to the file. Same when finished.
As for having multiple processes running, we might explicitly say 'Importing' when it's just one process and show process count only when there's multiple. Let me know, and I can better define this in the document.

**Reply by kgodey** [reply]
@ghislaineguerin I'm not sure that I understand the questions you're asking. I think you're asking whether it's okay to have the notification have different states based on the number of background processes happening (i.e. different states for one vs. more than one). If that's what you're asking, I think that's fine.

Let me know if that doesn't answer your question or if you have more questions.

**Reply by ghislaineguerin** [reply]
@kgodey yes, that was my question. Thanks

**Reply by pavish** [reply]
@ghislaineguerin We can change how the indicator looks based on different states. I think that should be okay. However, we would need a dropdown, next to the indicator, which opens the list of all async operations.

The dropdown would be needed, even if it's a single file, because it could belong to a different schema, and we need to be able to view the details.

If multiple files are being imported across schemas, and we change schemas/db, we would still need to see the file details, and click on the individual files on the dropdown list to move to the respective table.

**Reply by ghislaineguerin** [reply]
@pavish 
Yes, it would have a dropdown so that all operations are visible, I'm going to include that in the prototype. Good point on the potential scenario of users navigating to a different schema, I will incorporate that as well.

**Comment by ghislaineguerin** [comment]
@kgodey @pavish 
Review notes have been added to the document and listed below for reference.

## Global Notifications
Global notifications are needed to help users keep track of ongoing operations and navigate back to the particular pages. The location of these notifications should be a permanent one that is always visible to users, especially in scenarios such as moving from one database or schema to another.

## Notification States
The content and layout of these notifications might change dynamically according to the number or type of operations. In a single operation, the notification might show specific details, such as the name of the file being imported. Once more operations of the same type are added, it might change to offer a total count rather than details.
When operations are of a distinct type, the design might reduce the notification to only showing the number of operations, and users will find the actual type information in the dropdown.

**Reply by kgodey** [reply]
@ghislaineguerin I looked through the finished spec. It looks good, but I have a couple of questions/concerns:
- Where are we telling the users that they shouldn't navigate away from the browser?
- Where is the behavior of content and layout changes of notification detailed?

**Reply by ghislaineguerin** [reply]
@kgodey I've updated the document to include details for both. There's a new design file linked under notifications.

**Reply by kgodey** [reply]
Thanks @ghislaineguerin! Looks good.
